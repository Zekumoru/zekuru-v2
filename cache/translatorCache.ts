import { AuthorizationError, Translator, TranslatorOptions } from 'deepl-node';
import { Collection } from 'discord.js';
import GuildKey, { IGuildKey } from '../models/GuildKey';
import { loadLanguages } from '../translation/languages';
import { errorDebug } from '../utils/logger';
import { decrypt, encrypt } from '../utils/crypt';

export const translatorOptions: TranslatorOptions = {
  appInfo: {
    appName: 'Zekuru-v2 Demo',
    appVersion: '1.0.0',
  },
  minTimeout: 500, // 500 ms
  maxRetries: 10,
};

const cacheTranslators = new Collection<string, Translator>();

const fetchOrCreateGuildKey = async (guildId: string, hashedApiKey: string) => {
  // fetch from db, if not exists, create new
  const guildKey = await GuildKey.findOne({ id: guildId });
  if (guildKey) return [guildKey, false] as const;

  // save api key to db
  const newGuildKey = new GuildKey({
    id: guildId,
    key: hashedApiKey,
  });
  await newGuildKey.save();

  return [newGuildKey, true] as const;
};

const set = async (guildId: string, apiKey: string) => {
  // create translator client
  const translator = new Translator(apiKey, translatorOptions);

  // this should throw error if invalid api key
  await translator.getUsage();
  await loadLanguages(translator);

  // encrypt api key
  const encryptedApiKey = encrypt(apiKey);

  // update in db
  const [guildKey, isNewlyCreated] = await fetchOrCreateGuildKey(
    guildId,
    encryptedApiKey
  );

  if (!isNewlyCreated) {
    guildKey.overwrite({
      id: guildId,
      key: encryptedApiKey,
      createdAt: guildKey.createdAt,
    } as IGuildKey);
    await guildKey.save();
  }

  // update cache
  cacheTranslators.set(guildId, translator);
};

const get = async (guildId: string) => {
  // get from cache
  const translatorCache = cacheTranslators.get(guildId);
  if (translatorCache) return translatorCache;

  // if not exists in cache, fetch from api key from db
  // then create new translator and set to cache
  const guildKey = await GuildKey.findOne({ id: guildId });
  if (!guildKey) return;

  try {
    const apiKey = decrypt(guildKey.key);
    const translator = new Translator(apiKey, translatorOptions);

    // load languages
    await loadLanguages(translator);

    // update cache
    cacheTranslators.set(guildId, translator);
    return translator;
  } catch (error) {
    // delete invalid api key from db and "sign out" user by removing it from cache
    if (error instanceof AuthorizationError) {
      await unset(guildId);
    }
    errorDebug(error);
  }
};

const unset = async (guildId: string) => {
  // check if it exists first
  if ((await get(guildId)) == null) return;

  // delete from db
  await GuildKey.deleteOne({ id: guildId });

  // remove from cache
  cacheTranslators.delete(guildId);
};

export default { set, unset, get };
