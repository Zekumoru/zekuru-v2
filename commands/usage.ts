import { ChatInputCommandInteraction, SlashCommandBuilder } from 'discord.js';
import { createCommand } from '../types/DiscordCommand';
import translatorCache from '../cache/translatorCache';

const data = new SlashCommandBuilder()
  .setName('usage')
  .setDescription('Shows current usage and remaining characters.');

const execute = async (interaction: ChatInputCommandInteraction) => {
  // check if this bot has a translator
  if (!interaction.guildId) return;

  const translator = await translatorCache.get(interaction.guildId);
  if (translator == null) {
    interaction.reply({
      content: `Cannot get usage, no api key found. Please sign in using the \`sign-in\` command.`,
    });
    return;
  }

  const usage = await translator.getUsage();
  if (!usage.character) {
    await interaction.reply('An error occurred. Could not get usage.');
    return;
  }

  if (usage.character.limitReached()) {
    await interaction.reply('Usage limit has been reached!');
    return;
  }

  const { count, limit } = usage.character;
  const remaining = limit - count;
  await interaction.reply(
    `Usage count: **${count.toLocaleString('en-US')} / ${limit.toLocaleString(
      'en-US'
    )} characters**; Remaining: **${remaining.toLocaleString(
      'en-US'
    )} characters**.`
  );
};

export default createCommand({
  cooldown: 5,
  data,
  execute,
});
