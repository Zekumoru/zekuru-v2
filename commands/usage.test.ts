import usage from './usage';
import translatorCache from '../cache/translatorCache';
import createMockInteraction from '../test/createMockChatInputCommandInteraction';

jest.mock('../cache/translatorCache.ts');

describe('/usage command', () => {
  const mockTranslatorCache = jest.mocked(translatorCache);

  it(`should notify user if they're not signed in`, async () => {
    const interaction = createMockInteraction();
    mockTranslatorCache.get.mockResolvedValue(null as any);

    await usage.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should notify user if cannot get usage`, async () => {
    const interaction = createMockInteraction();
    mockTranslatorCache.get.mockResolvedValue({ getUsage: () => ({}) } as any);

    await usage.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should notify user if limit reached`, async () => {
    const interaction = createMockInteraction();
    mockTranslatorCache.get.mockResolvedValue({
      getUsage: () => ({
        character: {
          limitReached: () => true,
        },
      }),
    } as any);

    await usage.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should show usage`, async () => {
    const interaction = createMockInteraction();
    mockTranslatorCache.get.mockResolvedValue({
      getUsage: () => ({
        character: {
          count: 56789,
          limit: 500000,
          limitReached: () => false,
        },
      }),
    } as any);

    await usage.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });
});
