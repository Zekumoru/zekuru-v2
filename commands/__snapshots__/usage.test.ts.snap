// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/usage command should notify user if cannot get usage 1`] = `"An error occurred. Could not get usage."`;

exports[`/usage command should notify user if limit reached 1`] = `"Usage limit has been reached!"`;

exports[`/usage command should notify user if they're not signed in 1`] = `
{
  "content": "Cannot get usage, no api key found. Please sign in using the \`sign-in\` command.",
}
`;

exports[`/usage command should show usage 1`] = `"Usage count: **56,789 / 500,000 characters**; Remaining: **443,211 characters**."`;
